src/mlir_build.cpp

```diff
@@ -122,4 +136,14 @@ boost::filesystem::path MlirBuildViaHip(boost::optional<TmpDir>& tmp_dir,
     return HipBuild(tmp_dir, hip_filename, src, hip_options, target, true);
 }
 
+void MiirGenLaunchParams(const std::string& params, size_t& local_size, size_t& global_size)
+{
+    AutoMiirHandle handle(params.c_str());
+    miirLowerInit();
```

Comment:
[Q] I guess you need to pass handle to it?

Reply:
So far I don't see a good reason to make users of `mlir_build` understand the implementation detail like`AutoMiirHandle` (Therefore I hide the type in an anonymous namespace). In each use case of `MiirGenLaunchParams` or `MiirBuildViaHip`, there's no overhead I see so far.

Reply:
`miirLowerInit(handel());` ?

Reply:
Gotcha, I see what you mean now. No this initialization initialize the lowering pipeline and is handle independent.

Reply:
How it could be so? Where the internal state of the lowering pipeline resides?

Reply:
This is important Q.

Reply:
The lowering pipelines needs to be registered as LLVM passes before being invoked in actual lowering step, [here](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/miopen-dialect/mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp#L265-L274). Previously the `miirLowerBin` itself will do the initialization as well when it is called. But now since we added `miirLowerTuningParams` I'd rather maintain one orthogonal function that does the common initialization logic instead of duplicated code.

Reply:
Are you sure that https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/miopen-dialect/mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp#L265-L274 is thread-safe?

Reply:
It is: user is allowed to initialize as many times as they want. It won't do anything concrete if pipeline is already intialized.

Say we have thread `a` and thread `b`, as long as thread a/b do initialize before their own lowering, `LLVM` will be happy. It doesn't matter if the global sequence happen to be that the initialization happens in middle of the lowering.

Reply:
I will try to poke on it a bit more and see if I can skip those, say, if client is already initialized. Even if the whole thing can work  doesn't mean that it should be repeatedly initialized.

Reply:
What if thread A is doing the lowering but thread B is initializing? That is the case I am worrying of. Solution:
```C++
// C++
bool lazy_init()
{
    static const bool once = []()
        {
            ... // initialization
            return true;
        }();
    return once;
}
```
This can be done directly in the library.

Reply:
I checked around and find the documents [here](https://llvm.org/doxygen/classllvm_1_1PassRegistry.html#details). Looks like it confirmed that: Each thread has to do its own registration. This means that thread A and B should both do their own intialization before hand.

Very nice code snippet. I think I see this implementation in singleton pattern but never thought about applying this elsewhere. Will adopt it in library implementation.

Reply:
What this means is that I shouldn't expose the initialization as a client function, but as a convenient function inside the library. I shouldn't expect the client to be used in one way rather than another.

Reply:
> Will adopt it in library implementation.

Maybe. But if applied here it will perform initialization in one thread only.

> I checked around and find the documents [here](https://llvm.org/doxygen/classllvm_1_1PassRegistry.html#details). 

Well, it says "you will need to use a _separate_ PassRegistry on each thread" but this is not the same as _initialization of the same instance of the PassRegistry_ in each thread.

MIOpen spawns threads during build. In order to stay in the thread-safe realm, we need to create separate instance of PassRegistry per each thread. The simplest way to reach that goal is to aggregate PassRegistry in the MLIR handle (by value or by pointer/reference). Is it possible?

Reply:
> MIOpen spawns threads during build. In order to stay in the thread-safe realm, we need to create separate instance of PassRegistry per each thread. 

I think I got myself confused in the `PassRegistry` reading. `PassRegistry` and initialization are two independent topics and yes, we do have separate instance of `PassRegistry` per thread(call).

Refer to [here](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/miopen-dialect/mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp#L284).

Reply:
Ok, let's get back to the original question, https://github.com/ROCmSoftwarePlatform/MIOpen/pull/811#discussion_r599902994. Is it Ok to initialize by calling `miirLowerInit()` once? Would it work for all threads, all handles? If so, then let's implement lazy initialization, see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/811#discussion_r599941733.

This is important because I believe that [the sequence of initialization calls](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/miopen-dialect/mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp#L265-L274) in thread A should not happen in parallel with some other thread B that performs lowering.

Reply:
MLIR [issue](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/115) filed for this

Reply:
This is not a problem until we do not have > 1 MLIR solver in the library. Resolved.

Reply:
...and we still may have issues with multi-threaded user's applications, but not now because MLIR Solvers are OFF by default.